solidity/contracts/liquidity-protection/TokenTimeWeightedAverage.sol

```diff
@@ -0,0 +1,375 @@
+        return (s.n, s.d);
+    }
+
+    /**
+     * @dev returns whether a specific sample exists
+     *
+     * @param _token the token the data is accumulated for
+     * @param _time the sampling time
+     *
+     * @return whether the sample exists
+     */
+    function sampleExists(IERC20Token _token, uint256 _time)
+        external
+        view
+        override
+        validAddress(address(_token))
+        returns (bool)
+    {
+        TokenData storage tokenData = data[_token];
+
+        return tokenData.timestamps[_time];
+    }
+
+    /**
+     * @dev returns the first and the last sample times
+     *
+     * @param _token the token the data is accumulated for
+     *
+     * @return the first and the last sample times
+     */
+    function sampleRange(IERC20Token _token) external view override returns (uint256, uint256) {
+        TokenData memory tokenData = data[_token];
+
+        return (tokenData.firstSampleTime, tokenData.lastSampleTime);
+    }
+
+    /**
+     * @dev adds a new sample to the accumulator
+     * can only be called by a seeder
+     *
+     * @param _token the token the data is accumulated for
+     * @param _sample the sample to add
+     * @param _time the sampling timestamp
+     */
+    function addSample(
+        IERC20Token _token,
+        Fraction memory _sample,
+        uint256 _time
+    ) private {
+        TokenData storage tokenData = data[_token];
+        uint256 lastSampleTime = tokenData.lastSampleTime;
+
+        // make sure that the samples are added in the correct order
+        require(_time >= lastSampleTime, "ERR_WRONG_ORDER");
+
+        // if this is the initialization sample for this contract, set it as the first sample and return
+        if (tokenData.firstSampleTime == 0) {
+            tokenData.firstSampleTime = _time;
+            tokenData.lastSampleTime = _time;
+
+            tokenData.accumulators[_time] = _sample;
+            tokenData.timestamps[_time] = true;
+
+            emit SampleAdded(_token, _sample.n, _sample.d, _time);
+
+            return;
+        }
+
+        Fraction memory lastAccumulator = tokenData.accumulators[lastSampleTime];
+
+        // update the previous accumulator value once per-block. we would need it in order to accumulate
+        // same-block changes
+        if (_time > tokenData.prevAccumulatorUpdateTime) {
+            tokenData.prevAccumulatorUpdateTime = _time;
+            tokenData.prevAccumulator = lastAccumulator;
+        }
+
+        // if we already have a sample for this timestamp - use the backup of the previous accumulator and use it in
+        // combination with the new sample
+        if (_time == lastSampleTime) {
+            lastAccumulator = tokenData.prevAccumulator;
+        }
+
+        // accumulate the current value in combination with the last accumulator using the TWA formuls:
+        //  ACC[time] = ACC[time - timeDiff] + value * timeDiff
```

Comment:
I think that you should change this comment to:
```
// accumulator = lastAccumulator + smaple * timeDelta
```
If you don't think so, then just note the extra space after the slashes.

Reply:
Sure, why not?

Reply:
Only reason not to is because I'm wrong.